File: UNIT-01/orchestrator/session.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/charmbracelet/huh"
	"github.com/charmbracelet/lipgloss"
)

type SessionMeta struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	LastUpdated time.Time `json:"last_updated"`
}

type SessionData struct {
	SessionMeta
	History []Message `json:"history"`
}

func getSessionsDir() string {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".ruthen", "unit01", "sessions")
	os.MkdirAll(dir, 0755)
	return dir
}

func SaveSession(id string, name string, history *History) error {
	if id == "" {
		return nil
	}
	dir := getSessionsDir()
	path := filepath.Join(dir, id+".json")

	data := SessionData{
		SessionMeta: SessionMeta{
			ID:          id,
			Name:        name,
			LastUpdated: time.Now(),
		},
		History: history.messages,
	}

	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, b, 0644)
}

func LoadSession(id string, history *History) (string, error) {
	dir := getSessionsDir()
	path := filepath.Join(dir, id+".json")

	b, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}

	var data SessionData
	if err := json.Unmarshal(b, &data); err != nil {
		return "", err
	}

	history.messages = data.History
	return data.Name, nil
}

func HandleSessionsMenu(history *History) string {
	dir := getSessionsDir()
	entries, err := os.ReadDir(dir)
	if err != nil {
		fmt.Println("No sessions found.")
		return ""
	}

	var sessions []SessionMeta
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".json") {
			b, err := os.ReadFile(filepath.Join(dir, e.Name()))
			if err == nil {
				var meta SessionMeta
				if json.Unmarshal(b, &meta) == nil {
					sessions = append(sessions, meta)
				}
			}
		}
	}

	sort.Slice(sessions, func(i, j int) bool {
		return sessions[i].LastUpdated.After(sessions[j].LastUpdated)
	})

	var options []huh.Option[string]
	options = append(options, huh.NewOption(lipgloss.NewStyle().Foreground(BrandColor).Render("+ Start New Session"), "new"))

	for _, s := range sessions {
		timeStr := s.LastUpdated.Format("Jan 02, 3:04 PM")
		label := fmt.Sprintf("▣ %-30s (%s)", s.Name, timeStr)
		options = append(options, huh.NewOption(label, s.ID))
	}

	var selectedID string
	formErr := huh.NewSelect[string]().
		Title("Resume a Past Session").
		Options(options...).
		Value(&selectedID).
		Run()

	if formErr != nil {
		return ""
	}

	return selectedID
}

File: UNIT-01/orchestrator/src/session.rs
use crate::history::History;
use crate::types::Message;
use chrono::Utc;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct SessionMeta {
    pub id: String,
    pub name: String,
    pub last_updated: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SessionData {
    #[serde(flatten)]
    pub meta: SessionMeta,
    pub history: Vec<Message>,
}

fn sessions_dir() -> String {
    if let Some(home) = home::home_dir() {
        let dir = format!("{}/.ruthen/unit01/sessions", home.display());
        let _ = std::fs::create_dir_all(&dir);
        dir
    } else {
        "/tmp/ruthen-sessions".to_string()
    }
}

pub fn save_session(id: &str, name: &str, history: &History) -> Result<(), String> {
    if id.is_empty() {
        return Ok(());
    }
    let dir = sessions_dir();
    let path = format!("{}/{}.json", dir, id);

    let data = SessionData {
        meta: SessionMeta {
            id: id.to_string(),
            name: name.to_string(),
            last_updated: Utc::now().to_rfc3339(),
        },
        history: history.messages.clone(),
    };

    let json = serde_json::to_string_pretty(&data).map_err(|e| e.to_string())?;
    std::fs::write(&path, &json).map_err(|e| e.to_string())?;
    Ok(())
}

pub fn load_session(id: &str, history: &mut History) -> Result<String, String> {
    let dir = sessions_dir();
    let path = format!("{}/{}.json", dir, id);

    let data = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
    let session: SessionData = serde_json::from_str(&data).map_err(|e| e.to_string())?;

    history.messages = session.history;
    Ok(session.meta.name)
}

pub fn list_sessions() -> Vec<SessionMeta> {
    let dir = sessions_dir();
    let mut sessions = Vec::new();

    if let Ok(entries) = std::fs::read_dir(&dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("json") {
                if let Ok(data) = std::fs::read_to_string(&path) {
                    if let Ok(meta) = serde_json::from_str::<SessionMeta>(&data) {
                        sessions.push(meta);
                    }
                }
            }
        }
    }

    sessions.sort_by(|a, b| b.last_updated.cmp(&a.last_updated));
    sessions
}
